lib/pacemaker/pcmk_sched_instances.c

```diff
@@ -394,79 +394,88 @@ pcmk__assign_instances(pe_resource_t *collective, GList *instances,
+    if (instance->children != NULL) {
+        for (iter = instance->children; iter != NULL; iter = iter->next) {
+            check_instance_state((const pe_resource_t *) iter->data, state);
         }
-
         return;
     }
 
-    CRM_ASSERT(active != NULL);
-    CRM_ASSERT(starting != NULL);
-    CRM_ASSERT(stopping != NULL);
-
-    if (rsc->running_on) {
-        *active = TRUE;
+    if (instance->running_on != NULL) {
+        instance_state |= instance_active;
     }
 
-    gIter = rsc->actions;
-    for (; gIter != NULL; gIter = gIter->next) {
-        pe_action_t *action = (pe_action_t *) gIter->data;
-
-        if (*starting && *stopping) {
-            return;
-
-        } else if (pcmk_is_set(action->flags, pe_action_optional)) {
-            pe_rsc_trace(rsc, "Skipping optional: %s", action->uuid);
-            continue;
-
-        } else if (!pcmk_any_flags_set(action->flags,
-                                       pe_action_pseudo|pe_action_runnable)) {
-            pe_rsc_trace(rsc, "Skipping unrunnable: %s", action->uuid);
-            continue;
-
-        } else if (pcmk__str_eq(RSC_STOP, action->task, pcmk__str_casei)) {
-            pe_rsc_trace(rsc, "Stopping due to: %s", action->uuid);
-            *stopping = TRUE;
-
-        } else if (pcmk__str_eq(RSC_START, action->task, pcmk__str_casei)) {
-            if (!pcmk_is_set(action->flags, pe_action_runnable)) {
-                pe_rsc_trace(rsc, "Skipping pseudo-op: %s run=%d, pseudo=%d",
-                             action->uuid,
-                             pcmk_is_set(action->flags, pe_action_runnable),
-                             pcmk_is_set(action->flags, pe_action_pseudo));
+    // Check each of the instance's actions for runnable start or stop
+    for (iter = instance->actions;
+         (iter != NULL) && !pcmk_all_flags_set(instance_state,
+                                               instance_starting
+                                               |instance_stopping);
+         iter = iter->next) {
+
+        const pe_action_t *action = (const pe_action_t *) iter->data;
+
+        if (pcmk__str_eq(RSC_START, action->task, pcmk__str_none)) {
+            if (!pcmk_is_set(action->flags, pe_action_optional)
+                && pcmk_is_set(action->flags, pe_action_runnable)) {
+                pe_rsc_trace(instance, "Instance is starting due to %s",
+                             action->uuid);
+                instance_state |= instance_starting;
+            } else {
+                pe_rsc_trace(instance,
+                             "%s doesn't affect %s state (unrunnable)",
+                             action->uuid, instance->id);
+            }
+
+        } else if (pcmk__str_eq(RSC_STOP, action->task, pcmk__str_none)) {
+            if (!pcmk_is_set(action->flags, pe_action_optional)
+                && pcmk_any_flags_set(action->flags,
+                                      // Pseudo-stops are implied by fencing
```

Comment:
If the instance is a group, it will fall into the first loop checking the state of all the children, which means this code can only be reached for primitives. For primitives, there is no such thing as a pseudo-start, and yes pseudo-stops are only those implied by fencing. If we're fencing a node, we don't want to actually try to execute stop actions on it, so we convert all stop actions to pseudo-actions and order them after fencing. I can improve the comments.